File: cielo24_Ruby/cielo24_gem/lib/cielo24/web_utils.rb
module Cielo24
  class WebUtils

    require 'json'
    require 'httpclient'
    require 'timeout'
    include JSON

    BASIC_TIMEOUT = 60     # seconds
    DOWNLOAD_TIMEOUT = 300 # seconds
    @@LAST_URL = "none" # For logging purposes

    def self.LAST_URL
      return @@LAST_URL
    end

    def self.get_json(uri, method, timeout, query=nil, headers=nil, body=nil)
      response = http_request(uri, method, timeout, query, headers, body)
      return JSON.parse(response)
    end

    def self.http_request(uri, method, timeout, query=nil, headers=nil, body=nil)
      http_client = HTTPClient.new
      http_client.cookie_manager = nil
      http_client.send_timeout = 60*60*24*7 # HTTPClient default timeout set to 7 days, our own timeout handler is down below
      @@LAST_URL = uri + (query.nil? ? "" : "?" + URI.encode_www_form(query))

      # Timeout block:
      begin
        # Error is raised if the following block fails to execute in 'timeout' sec:
        Timeout.timeout(timeout) { # nil timeout = infinite

          response = http_client.request(method, uri, query, body, headers, nil)
          # Handle web errors
          if response.status_code == 200 or response.status_code == 204
            return response.body
          else
            json = JSON.parse(response.body)
            raise WebError.new(json["ErrorType"], json["ErrorComment"])
          end

        }
      rescue Timeout::Error
        raise TimeoutError.new("The HTTP session has timed out.")
      end
    end
  end

  class WebError < StandardError
    attr_reader :type
    def initialize(type, comment)
      super(comment)
      @type = type
    end

    def to_s
      return @type + " - " + super.to_s
    end
  end

  class TimeoutError < StandardError
    def initialize(message)
      super(message)
    end
  end
end

File: cielo24_PHP/src/cielo24/WebUtils.php
<?php

use GuzzleHttp\Client;
use GuzzleHttp\Stream\Stream;
use GuzzleHttp\Exception\RequestException;

class WebUtils {
    const BASIC_TIMEOUT = 60;           # seconds
    const DOWNLOAD_TIMEOUT = 300;       # seconds
    const UPLOAD_TIMEOUT = 604800;      # seconds (7 days)

    public static function getJson($base_uri, $path, $method, $timeout, $query=array(), $headers=array(), $body=null) {
        $response = WebUtils::httpRequest($base_uri, $path, $method, $timeout, $query, $headers, $body);
        return json_decode($response, true);
    }

    public static function httpRequest($base_uri, $path, $method, $timeout, $query=array(), $headers=array(), $body=null) {
        if ($query == null){
            $query = array();
        }
        if ($headers == null) {
            $headers = array();
        }

        $url = $base_uri . $path;
        if (count($query) > 0) {
            $url .= "?" . http_build_query($query);
        }

        $http_client = new Client();
        $http_request = $http_client->createRequest($method, $url,["timeout" => $timeout]);
        $http_request->addHeaders($headers);

        if ($body != null) {
            $http_request->setBody(Stream::factory($body));
        }

        try {
            $response = $http_client->send($http_request);
            return $response->getBody();
        } catch(RequestException $e) {
            $json = json_decode($e->getResponse()->getBody(), true);
            throw new WebError($json["ErrorType"], $json["ErrorComment"]);
        }
    }
}

class WebError extends Exception {

    public $errorType;
    public $errorComment;

    public function __construct($type, $comment) {
        parent::__construct();
        $this->errorType = $type;
        $this->errorComment = $comment;
    }

    public function __toString() {
        return $this->errorType . " - " . $this->errorComment;
    }
}